src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultHtmlFactsService.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    internal class DefaultHtmlFactsService : HtmlFactsService
+    {
+        public override bool TryGetElementInfo(SyntaxNode element, out SyntaxToken containingTagNameToken, out SyntaxList<RazorSyntaxNode> attributeNodes)
+        {
+            if (element is MarkupStartTagSyntax startTag)
+            {
+                containingTagNameToken = startTag.Name;
+                attributeNodes = startTag.Attributes;
+                return true;
+            }
+
+            if (element is MarkupTagHelperStartTagSyntax startTagHelper)
+            {
+                containingTagNameToken = startTagHelper.Name;
+                attributeNodes = startTagHelper.Attributes;
+                return true;
+            }
+
+            containingTagNameToken = null;
+            attributeNodes = default;
+            return false;
+        }
+
+        public override bool TryGetAttributeInfo(SyntaxNode attribute, out SyntaxToken containingTagNameToken, out string selectedAttributeName, out SyntaxList<RazorSyntaxNode> attributeNodes)
+        {
+            if ((attribute is MarkupMiscAttributeContentSyntax ||
+                attribute is MarkupMinimizedAttributeBlockSyntax ||
+                attribute is MarkupAttributeBlockSyntax ||
+                attribute is MarkupTagHelperAttributeSyntax ||
+                attribute is MarkupMinimizedTagHelperAttributeSyntax ||
+                attribute is MarkupTagHelperDirectiveAttributeSyntax ||
+                attribute is MarkupMinimizedTagHelperDirectiveAttributeSyntax) &&
+                TryGetElementInfo(attribute.Parent, out containingTagNameToken, out attributeNodes))
+            {
+                if (attribute is MarkupTagHelperAttributeSyntax)
```

Comment:
I remembered that I DID make a change here: previously selectedAttributeName under no circumstances returned a value, and I made it return a value in this case. This broke a test that expected completions not to supply the things they hover over, but it turns out that that's how everything else works anyway.

The question though becomes: should we return a value in a wider case than `MarkupTagHelperAttributeSyntax`?

Reply:
Also, how was this list of Syntax's arrived at?